### PR TITLE
Custom maneuver behaviour gadget (Resolves: #75)

### PR DESCRIPTION
**Pull request by chocolatemalc**
### Work done
This is a new lua gadget that modifies the default behaviour of the "Maneuver" movestate.

#### OLD BEHAVIOUR
1. IDLE "Maneuver" state enabled unit encounters enemy unit in sight range.
2. "Maneuver" unit issues an attack command on enemy unit.
3. When no enemy units remain in sight range, "Maneuver" unit stops.

#### NEW BEHAVIOUR
1. IDLE "Maneuver" state enabled unit encounters enemy unit in sight range.
2. "Maneuver" unit issues an attack command on enemy unit.
3. When no enemy units remain in sight range, "Maneuver" returns to the coordinates of the last MOVE (or FIGHT) command it received prior to being IDLE.

I've written this with the following in mind:
1. Do not override behaviours other than MANEUVER. 
 Here it is worth mentioning that during testing I found that ATTACK commands issued in the RETURNFIRE firestate are not reset in the same way by the engine as ATTACK commands in the FIREATWILL firestate. RETURNFIRE ATTACK commands persist beyond sight range. FIREATWILL ATTACK commands will be canceled outside of sight range. I do not know if this is intended, and I have not modified the behaviour in this gadget, even though it  produces unintuitive in the case where you are using RETURNFIRE + MANEUVER (I'd say an extreme edge case) 
1. Do no apply to air units and new behaviour ignores air units.

#### Addresses Issue(s)
beyond-all-reason/Beyond-All-Reason#75

#### Testing
- I've attempted to test every NEW BEHAVIOUR scenario I could think of that to ensure it's functioning as described above (including while moving, issuing other commands, stopping, vs Air, etc).
- I have not tested any navy unit
- I have only tested the OLD BEHAVIOUR in the simple IDLE "Manuever" land unit vs enemy land unit entering sight range.

### Questions
For my own education and for best performance I would appreciate any feedback, particularly on the following (comments exist in initial version of the PR):

``` lua
local UPDATE_INTERVAL = 15 --is this too much/too little?
```

``` lua
--first time trying something like this. is it necessary or common to distribute logic across periods of frames?
if unitID % UPDATE_INTERVAL == frame % UPDATE_INTERVAL then 
```

``` lua
                         if validTargetID then
                            Spring.Echo(validTargetID, "is Valid target")
                            -- Spring.Echo(cmdQueue)
                            if unitStates.firestate == FIRESTATE_RETURNFIRE then
                                local attackedRecently = recentlyAttacked[unitID] and (frame - recentlyAttacked[unitID] <= ATTACK_MEMORY_DURATION)
                                if attackedRecently then
                                    if not cmdQueue[1] then
                                        -- Spring.Echo("[RETURN-FIRE CHASE]", unitID, "->" ,validTargetID) --couldn't get this echo to work
                                        Spring.Echo("[RETURN-FIRE CHASE]")
                                        spGiveOrderToUnit(unitID, CMD_ATTACK, {validTargetID}, {})
                                    end
                                end
                            else
                                if not cmdQueue[1] then
                                    --This block never runs with current settings (UPDATE_INTERVAL = 15). Race condition with engine MOVESTATE_MANUEVER?
                                    -- Spring.Echo("[CHASE]", unitID, "->", validTargetID)--couldn't get this echo to work
                                    Spring.Echo("[CHASE]") 
                                    spGiveOrderToUnit(unitID, CMD_ATTACK, {validTargetID}, {})
                                end
                            end
```
P.S. I know there's a little repetition that I can refactor. I still plan on doing this, but it's very late and I'm too excited to share this PR to wait.

### CODE REVIEW

**Comment by SethDGamre** [comment]
Marks for game design team approval because currently the existence of maneuver is not in alignment with the game design document as written. 
Screenshot_2025-07-10-10-23-33-63_df198e732186825c8df26e3c5a10d7cd

Sounds like unimprovement over the default behavior in the interim however to me.

**Comment by SethDGamre** [comment]
@beyond-all-reason/game-design-team

**Comment by SethDGamre** [comment]
Oh, actually now that I read it with my eyeballs, this is exactly what we need. 
It is Roam that needs to go away.
Whoops.

**Comment by efrec** [comment]
A couple notes on our lua code style:

**Use tabs, not spaces, to indent.**

The file uses mostly spaces. Change these to tabs, when you can.

**Don't avoid using newlines.**

Always place them before and after closures like if/then blocks, like so:

```lua
local unitStates = spGetUnitStates(unitID)

if unitStates.firestate ~= FIRESTATE_HOLDFIRE then
    -- do whatever
end

Spring.Echo("do whatever else")
```

Except when they are immediately nested:

```lua
if unitStates.firestate ~= FIRESTATE_HOLDFIRE then
    if unitStates.movestate == MOVESTATE_MANEUVER then
        -- do whatever v2.0
    end
end
```

Once your code is spaced this way with newlines, it tends to make sense where else they might belong. But you don't need to put them anywhere else. It's up to you.

**Comment by efrec** [comment]
Units don't seek out targets only on sight radius. They have a search radius given by, unfortunately, more than one thing, like the maneuver leash radius. Long-range units will chase beyond their sight radius and into radar-only zones.

You may be better off finding out when a unit has idled, keeping that position, and restoring it to that position when you detect it idle for a second time (after it gets done chasing).

Also, I know you'll refactor this a bit. I'd say to go ahead, with the above in mind.

As for your questions:

1. The timeout for chasing a unit that disappears is very short, just a frames. You'd never want to update faster than maybe 5 frames. A delay of over a second might be frustrating or could allow peeling units away strategically with un-catch-able fodder. Somewhere in the middle seems fine.
2. This is a good way to do it. It's good stage effects to have each unit "think" on its own.
3. Not sure if there is a third question.

File: luarules/gadgets/unit_maneuver.lua [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+local spGetUnitDefID = Spring.GetUnitDefID
+local spGetGameFrame = Spring.GetGameFrame
+
+local unitSightDistances = {}
+local lastMovePosition = {}
+local recentlyAttacked = {}
+local nonAirUnits = {}
+
+function gadget:AllowCommand(unitID, unitDefID, unitTeam, cmdID, cmdParams, cmdOptions)
+        if (cmdID == CMD_MOVE or cmdID == CMD_FIGHT) and cmdParams[1] and cmdParams[2] and cmdParams[3] then
```
Comment (saurtron) [inline luarules/gadgets/unit_maneuver.lua]: Should `gadgetHandler:RegisterAllowCommand(cmdID)` at gadget:Initialize for the relevant cmdIDs

File: luarules/gadgets/unit_maneuver.lua [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+            lastMovePosition[unitID] = {
+                x = cmdParams[1],
```
Comment (saurtron) [inline luarules/gadgets/unit_maneuver.lua]: Should use tabs instead of spaces.

File: luarules/gadgets/unit_maneuver.lua [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+                y = y, 
+                z = z 
+            }
+        end
+    end
+end
+
+function gadget:UnitDamaged(unitID, unitDefID, unitTeam, damage, paralyzer, weaponID, attackerID, attackerDefID, attackerTeam)
+    if attackerID and attackerTeam ~= unitTeam then
+        recentlyAttacked[unitID] = spGetGameFrame
```
Comment (saurtron) [inline luarules/gadgets/unit_maneuver.lua]: Missing `()` -> `spGetGameFrame()`

File: luarules/gadgets/unit_maneuver.lua [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+                                    origPosition.x,
+                                    origPosition.y,
+                                    origPosition.z
+                                }, {})
+                            end
+                        end
+                    else
+                        local x, y, z = spGetUnitPosition(unitID)
+                        if x and y and z then
+                            lastMovePosition[unitID] = {
```
Comment (saurtron) [inline luarules/gadgets/unit_maneuver.lua]: would be better to define lastMovePosition as {x, y, z} instead of {x=x, y=y, z=z}, that way you avoid having to remake the table at spGiveOrderToUnit, also it's going to be slightly more memory efficient and I think easy to understand whats going on anyways.

File: luarules/gadgets/unit_maneuver.lua [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+                    z = z 
+                }
+            end
+        end
+    return true
+end
+
+function gadget:UnitCreated(unitID, unitDefID, unitTeam)
+    local def = UnitDefs[unitDefID]
+    if def and def.sightDistance and not def.canFly then
```
Comment (saurtron) [inline luarules/gadgets/unit_maneuver.lua]: Should precompute a table of UnitDefs that will pass the test instead of looking at UnitDefs all the time since that's quite slow.

Goal is to be able to test like this:

`if notFlyingWithSight[unitDefID] then`

(feel free to choose your own table name, I just wrote the first thing I could think of)

File: luarules/gadgets/unit_maneuver.lua [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+function gadget:UnitDestroyed(unitID)
+    unitSightDistances[unitID] = nil
+    lastMovePosition[unitID] = nil
+    recentlyAttacked[unitID] = nil
+    nonAirUnits[unitID] = nil
+end
+
+function gadget:GameFrame(frame)
+    for unitID, sightDist in pairs(unitSightDistances) do
+        if nonAirUnits[unitID] then
```
Comment (saurtron) [inline luarules/gadgets/unit_maneuver.lua]: what's the point of this check? Afaics all units in `unitSightDistances` are also at `nonAirUnits`.

File: luarules/gadgets/unit_maneuver.lua [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+                                        Spring.Echo("[RETURN-FIRE CHASE]")
+                                        spGiveOrderToUnit(unitID, CMD_ATTACK, {validTargetID}, {})
+                                    end
+                                end
+                            else
+                                if not cmdQueue[1] then
+                                    --This block never runs with current settings (UPDATE_INTERVAL = 15). Race condition with engine MOVESTATE_MANUEVER?
+                                    -- Spring.Echo("[CHASE]", unitID, "->", validTargetID)--couldn't get this echo to work
+                                    Spring.Echo("[CHASE]") 
+                                    spGiveOrderToUnit(unitID, CMD_ATTACK, {validTargetID}, {})
```
Comment (saurtron) [inline luarules/gadgets/unit_maneuver.lua]: No need for the tables, when having just one order parameter you can write it without table, also when no options, can just skip them, so:

```diff
-spGiveOrderToUnit(unitID, CMD_ATTACK, {validTargetID}, {})
+spGiveOrderToUnit(unitID, CMD_ATTACK, validTargetID)
```

**Review by saurtron — CHANGES_REQUESTED** [review]
Did a quick review and noted some issues.

Also I'd consider making this a widget instead of a gadget as it would probably be lighter that way (as every player processes just their units that way), and also ppl can opt-out of the behaviour or have alternative implementations easier.

Anyways you might want to wait for GDT approval before commiting more work to this, but I thought the comments will help you in other PRs too.

**Comment by chocolatemalc** [comment]
Thank you for the feedback. Really appreciate it. I'll have time later this weekend to follow up.

**Comment by chocolatemalc** [comment]
> A couple notes on our lua code style:
> 
> **Use tabs, not spaces, to indent.**
> 
> The file uses mostly spaces. Change these to tabs, when you can.
> 
> **Don't avoid using newlines.**
> 
> Always place them before and after closures like if/then blocks, like so:
> 
> ```lua
> local unitStates = spGetUnitStates(unitID)
> 
> if unitStates.firestate ~= FIRESTATE_HOLDFIRE then
>     -- do whatever
> end
> 
> Spring.Echo("do whatever else")
> ```
> 
> Except when they are immediately nested:
> 
> ```lua
> if unitStates.firestate ~= FIRESTATE_HOLDFIRE then
>     if unitStates.movestate == MOVESTATE_MANEUVER then
>         -- do whatever v2.0
>     end
> end
> ```
> 
> Once your code is spaced this way with newlines, it tends to make sense where else they might belong. But you don't need to put them anywhere else. It's up to you.

Thanks for the info. I attempted to replicate the format of other files I was using for reference. I've switched to tabs.

Can you clarify which newline format is preferred?
A:
``` lua
function gadget:UnitCreated(unitID, unitDefID, unitTeam)
	local def = UnitDefs[unitDefID]

	if def and def.sightDistance and not def.canFly then
		unitSightDistances[unitID] = def.sightDistance
		nonAirUnits[unitID] = true
	end

	if not lastMovePosition[unitID] then
		local x, y, z = spGetUnitPosition(unitID)
		
		if x and y and z then
			lastMovePosition[unitID] = { 
				x = x, 
				y = y, 
				z = z 
			}
		end

	end
	
end
```
B:
``` lua
function gadget:UnitCreated(unitID, unitDefID, unitTeam)
	local def = UnitDefs[unitDefID]

	if def and def.sightDistance and not def.canFly then
		unitSightDistances[unitID] = def.sightDistance
		nonAirUnits[unitID] = true
	end

	if not lastMovePosition[unitID] then
		local x, y, z = spGetUnitPosition(unitID)
		
		if x and y and z then
			lastMovePosition[unitID] = { 
				x = x, 
				y = y, 
				z = z 
			}
		end
	end
end
```

**Comment by efrec** [comment]
> Can you clarify which newline format is preferred?

Option B, for sure. Those are immediately nested.

**Comment by WatchTheFort** [comment]
This behaviour should share code with the idle builder repair functionality, as they implement similar behaviours

**Comment by sparr** [comment]
What is the next step for this PR? The implementation seems to at least mostly match what #75 was asking for. Is the problem that @chocolatemalc is gone and someone just needs to pick this up (hence the Needs Implementer label)?

**Comment by efrec** [comment]
See my longer comment on issue #75.

**Comment by chocolatemalc** [comment]
Hey there, I'm sorry haven't had much time to work on personal projects for a while now. Last time I was working on this, I was just working through the feedback. Primarily converting it from a gadget to a widget, which I was struggling with because the function I was using to read distance between units was not available in widget/client
